### DISCUSSION TITLE
[Question, mostly] Java::JavaAwt versus Java::Awt

### DISCUSSION THREAD

**Discussion by rubyFeedback** [jruby/jruby#8371] [Q&A] [answered]
I am playing around with jruby + swing right now, in particular writing a GUI agnostic
toolkit - a bit similar to glimmer. My idea is to abstract things into e. g. "Button.new"
being valid for all main use cases: ruby-gtk3 right now, hopefully ruby-tk eventually,
ruby-libui-ng, the web (via objectified HTML tags, e. g. able to treat HTML tags as
objects), perhaps ruby-SDL (via gosu) and so forth. Anyway.

Java + Swing may not be the most elegant toolkit and it is a bit messy too, but it
works out-of-the-box, which is great.

One thing that is a bit confusing is that we have AWT as well as swing on top of it.

The namespace we see here used for swing is:

    Java::JavaxSwing::JFrame
    Java::JavaxSwing

So, JavaxSwing.

For Awt it is:

    Java::JavaAwt::Frame
    Java::JavaAwt

In particular for Awt I found it a bit confusing that "Java" is repeated twice,
e. g.

    Java::JavaAwt

Why isn't it:

    Java::Awt

But looking at Swing, it is also a bit confusing:

    Java::JavaxSwing

Why is it Javax there? Why not Java? And why isn't it:

    Java::Swing

Now, I assume you guys had some kind of reasoning about this.

Perhaps:

    Java::Swing
    Java::Awt

may clash with something else. I actually had a name clash elsewhere,
where I had defined:

    Widgets::Frame

which conflicted with, I think, awt Frame (I think java_import may have 
pulled in some prior name or alias, e. g. if one were to do "include Java").

Anyway, I am mostly just curious about the naming choices here, not
necessarily meaning that anything should change. Has there been some
rationale about the name choices used? My initial attempt was to e. g.
try something like:

    Jawa::Awt::Frame

which did not work as it is instead:

    Java::JavaAwt::Frame

**Comment by enebo** [comment, marked as answer]
The first word ```Java``` is a place where we bind constants for Java classes.  The second word is the concatenated package name.  ```Java::JavaxSwing::JButton``` means a Java class JButton in the package javax.swing.  We cannot know all Java classes or where they load so we use a consistent unambiguous naming.  Java::Awt could be a Java class with no package which happens to be called Awt.

Most people tend to use ```java_import``` and do not see these constants unless you happen to notice the constant binding which is defined.

**Reply by headius** [reply]
Another detail here: We don't recommend the use of the `Java::JavaAwt...` form because it can be ambiguous when packages have caps in them, and as you found it's a bit of a guessing game to figure out that name. For clarity, it's better to use `Java::java.awt...` or `Java::java::awt` which both work and use the original package names. We probably will never remove the `JavaAwt` forms but their use is discouraged.

And the `java`, `javax`, `javafx`, `com`, and `org` packages are always available at the top level, so you can just reference `java.awt` from anywhere in JRuby.
